### ISSUE TITLE
Secured Actuator endpoints try to redirect to a login page rather than returning a 401

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#10907]
See https://github.com/spring-projects/spring-boot/issues/10306#issuecomment-341761010

**Comment by rwinch** [comment]
The behavior is quite intentional and I don't think it should change for the following reasons:

- Prompting for HTTP Basic in a browser is less secure because
  - There is no way to log out short of closing all the browser windows (and really who does this)
  - HTTP Basic has no time out like a session
  - HTTP Basic transfers a long term credential on every request increasing the likely hood of a long term credential leaking
- Using HTTP Basic takes 1/2 second longer per request. When passwords are properly stored (i.e. using BCrypt) validation of the password should take 1/2 a second. This means that if HTTP Basic is being used, the response time is increased by 1/2 second per request (the credentials are validated on every request)
- Requesting the endpoints with an accept type that does not include HTML will ensure that HTTP Basic is prompted. In a browser, users can append `.json` to the request URL to force HTTP Basic.
- Preemptive HTTP Basic is supported for any accept type. This means that automated tooling will still work.
- This is only the default behavior. The default security configuration is never going to be adequate for a real application, so we do expect users to modify this.
- If the client has an accept type that allows for HTML, then a redirect to an HTML page should be expected as it is a much better user experience.

**Comment by philwebb** [comment]
I'm very much against moving away from the Spring Security defaults and I think we should keep the existing behavior.

> In a browser, users can append `.json` to the request URL to force HTTP Basic.

This doesn't seem too onerous for the user, and given that preemptive HTTP Basic is supported for any accept type I don't see us breaking a lot of real API users.

**Comment by philwebb** [comment]
Actually it appears that adding `.json` as an extension isn't supported by actuator.

**Comment by dsyer** [comment]
I don't know if we need to change the new defaults, but I do think we need a way for hapless users (me) to become better informed about how things work. Maybe I'm looking in the wrong place, but it seems like the user guide doesn't reflect any of the above discussion (or the related discussion in #10306), and the frustration comes not when browsing a user guide, but in the trenches running apps in an IDE and being baffled by changes that are hard to understand.

Knowing that thing about "accept type that does not include HTML" could be very important. Better "deprecation" docs on existing 1.x configuration properties might also help, since some users will stumble on the changes in the IDE and configuration properties.

**Comment by wilkinsona** [comment]
> Actually it appears that adding .json as an extension isn't supported by actuator.

That's to avoid problems with the URL triggering unwanted content negotiation. See https://github.com/spring-projects/spring-boot/issues/8765 for details. If we want `.json` to work as @rwinch has described then I think we'd need to rework all of the actuator's endpoints to guarantee that the URLs path will never end with something that could be mistaken for a path extension.

**Comment by bclozel** [comment]
About using a `.json` extension to force content negotiation - we've removed that possibility from Spring WebFlux since this feature (path extension content negotiation strategy) can lead to security issues (including RFD attacks). These days, most clients support proper content negotiation, but we still provide a escape hatch with request parameters (see `ParameterContentNegotiationStrategy`).

@rwinch I agree that those defaults are quite sane.

> If the client has an accept type that allows for HTML, then a redirect to an HTML page should be expected as it is a much better user experience.

I worked recently on error handling with WebFlux in Boot and while being technically correct, this approach leads to the following:
* triggering an error in your app with a browser leads to a nice HTML error
* but triggering the same with a `curl` command also shows HTML in your console, since curl will send an `"Accept: */*"` header by default.

I ended up doing something like the following:
1. get the list of accepted media types from the request
2. remove `MediaType.ALL` from the list if present
3. then use HTML if one of the accepted media types is compatible with `text/html`, or use `application/json` by default

Do you think the Spring Security could consider such a change?
This would not change the sane defaults, but just align with the fact that browsers will send a `"text/html"` accepted media type, no matter what.

**Comment by rwinch** [comment]
@bclozel 

The algorithm I defined above is rather simplified. Spring Security should already ignore `Accept: */*` using [MediaTypeRequestMatcher.setIgnoredMediaTypes](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/web/util/matcher/MediaTypeRequestMatcher.html#setIgnoredMediaTypes-java.util.Set-) I verified with stock applications ([servlet](https://github.com/rwinch/spring-security-sample/tree/boot) / [webflux](https://github.com/rwinch/spring-security-sample/tree/boot-webflux)) that has Spring Security auto configuration that the result of curling it from the command line is a 401.

```bash
curl -v http://localhost:8080/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 401 Unauthorized
< WWW-Authenticate: Basic realm="Realm"
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Content-Type-Options: nosniff
< X-Frame-Options: DENY
< X-XSS-Protection: 1 ; mode=block
< content-length: 0
< 
* Connection #0 to host localhost left intact
```

Are you getting a different behavior or were you working of the basis of my incomplete definition of how Spring Security determines to return a 401? If you are getting something different, can you help me reproduce the problem as I think this should work.

Thanks!

**Comment by philwebb** [comment]
@rwinch 

I've tried with a M6 application generated from start.spring.io. Select `web`, `actuator` and `securirty` then start the application:

```
curl -v http://localhost:8080/application/env
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /application/env HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 302 
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Set-Cookie: JSESSIONID=DCF8A9C844323EF255E804CD70F3668E; Path=/; HttpOnly
< Location: http://localhost:8080/login
< Content-Length: 0
< Date: Wed, 15 Nov 2017 23:35:29 GMT
< 
```

The redirect is generated from `LoginUrlAuthenticationEntryPoint`

**Comment by rwinch** [comment]
Did you try Boot 2.0.0.BUILD.SNAPSHOT?

**Comment by philwebb** [comment]
Yeah, same:

```
curl -v http://localhost:8080/application/env
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /application/env HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 302 
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Set-Cookie: JSESSIONID=6F0213213F13066013F27EE3ABB99945; Path=/; HttpOnly
< Location: http://localhost:8080/login
< Content-Length: 0
< Date: Thu, 16 Nov 2017 00:44:49 GMT
< 
* Connection #0 to host localhost left intact
```

**Comment by rwinch** [comment]
Thanks for the details @philwebb It appears there is a problem w/ the standard servlet side. When I tested IntelliJ did not refresh (or report any errors) due to the fact that I still had the remote-shell starter which was not found. I created https://github.com/spring-projects/spring-security/issues/4831

Would anyone want to double check things behave as expected with WebFlux?

**Comment by wilkinsona** [comment]
I get a 401 with WebFlux:

```
curl -v localhost:8080/application/env
*   Trying ::1...
* Connected to localhost (::1) port 8080 (#0)
> GET /application/env HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.46.0
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< WWW-Authenticate: Basic realm="Realm"
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Content-Type-Options: nosniff
< X-Frame-Options: DENY
< X-XSS-Protection: 1 ; mode=block
< content-length: 0
<
* Connection #0 to host localhost left intact
```

**Comment by rwinch** [comment]
Thanks for confirming @wilkinsona 

It appears that the problem is related to the fact that in Spring Boot a Spring MVC  application uses `ContentNegotiationManager` [which ignores](https://github.com/spring-projects/spring-framework/blob/0fbfa643856b07eb0d952905e527765fbd5c274f/spring-web/src/main/java/org/springframework/web/accept/ContentNegotiationManager.java#L128) `MediaType.ALL`. This means that Spring Security is unable to use `MediaType.ALL` to trigger a 401 when `ContentNegotiationManager` is being used.

I could probably work around this using a different matcher, but at this point I think the problem is in `ContentNegotiationManager`. Thoughts?

**Comment by wilkinsona** [comment]
> It appears that the problem is related to the fact that in Spring Boot a Spring MVC application uses ContentNegotiationManager

Unless I've missed something, this isn't specific to Spring Boot. We're using the default `ContentNegotiationManager` configuration that's created by the `RequestMappingHandlerAdapter` that's created by `WebMvcConfigurationSupport`. It uses a single strategy: `HeaderContentNegotiationStrategy`.

> or Boot could provide a different ContentNegotiationStrategy implementation

If what I've said above is accurate, I don't think we should do this as it would mean we were deviating from Spring MVC's defaults.

> at this point I think the problem is in ContentNegotiationManager. Thoughts?

From what I've seen, I think the problem either lies in Spring MVC or Spring Security but I don't think I know enough to make a call one way or the other. Given that we're using Spring MVC's defaults I don't think the problem lies in Boot.

Perhaps @rstoyanchev or @bclozel can help us here?

**Comment by bclozel** [comment]
Yes, the `ContentNegotiationStrategy` has been like that for a while now (3.2.3), and I've traced back that particular part to [SPR-10513](https://jira.spring.io/browse/SPR-10513). I don't think we should change that as sending `Accept: */*` and sending nothing mean the same thing. Changing that behaviour would short-circuit other strategies in the list.

It looks like spring-projects/spring-security#4831 is now closed, but we still get this:

```
curl -v http://localhost:8080/actuator/env
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /actuator/env HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 302
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Set-Cookie: JSESSIONID=F13DF577D21C7178CB16FAFC991F426A; Path=/; HttpOnly
< Location: http://localhost:8080/login
< Content-Length: 0
< Date: Wed, 29 Nov 2017 16:53:16 GMT
<
* Connection #0 to host localhost left intact
```

@rwinch You were suggesting an issue in Spring Boot - does it mean you don't get that problem with plain Spring Framework?

Note that we're also currently reconsidering the path extension content negotiation in #11105 for MVC.

**Comment by rwinch** [comment]
@bclozel Spring Security is behaving correctly according to the configured `ContentNegotiationStrategy`. If `HeaderContentNegotiationStrategy` is used, then the application behaves as expected. It is surprising to me that `ContentNegotiationManager` resolves an empty `List<MediaType>` when it delegates to `HeaderContentNegotiationStrategy` which resolves `List<MediaType>` containing `MEDIA_TYPE_ALL`. 

Perhaps we should update the logic of `ContentNegotiationManager` to  be something like this:

```java
@Override
public List<MediaType> resolveMediaTypes(NativeWebRequest request) throws HttpMediaTypeNotAcceptableException {
	List<MediaType> defaultResult = Collections.emptyList();
	for (ContentNegotiationStrategy strategy : this.strategies) {
		List<MediaType> mediaTypes = strategy.resolveMediaTypes(request);
		if (mediaTypes.isEmpty()) {
			continue;
		} else if(mediaTypes.equals(MEDIA_TYPE_ALL)) {
			defaultResult = mediaTypes;
			continue;
		}
		return mediaTypes;
	}
	return defaultResult;
}
```

Thoughts?

**Comment by bclozel** [comment]
I've tried applying that change on master and ran the tests: nothing breaks.
I've also played with a couple of scenarios involving several strategies (headers, fixed and parameter) and everything checks.

It seems that in Spring Framework we rely on the fact that `"Accept:  "` and `"Accept: */*"` both mean "accept anything" on the receiving side and an empty produce condition will result in a "can produce `"*/*"`" (see `ProducesRequestCondition`).

I'm still not sure why we do have to make a change in those areas when both Framework and Security have been working like that for ages. I'm also confused about whether this is a bug in `ContentNegotiationManager` or not. If we do change that, it will return different results with `"Accept:  "` vs `"Accept: */*"`.

Also, looking at Spring Security's `MediaTypeRequestMatcher`, it looks a lot like `ContentNegotiationManager`, but inside out. I guess you could achieve the same thing with a `ContentNegotiationManager` that holds a `HeaderContentNegotiationStrategy` and falls back to a `FixedContentNegotiationStrategy`.

At that point, it's quite clear that this has nothing to do with Spring Boot and we should move this to a different place. @rstoyanchev, do you have anything in mind that could point to the right direction from here?

**Comment by rstoyanchev** [comment]
I don't quite understand why it works with WebFlux but not with Spring MVC. They have the exact same logic with regards to a strategy returning a list with a single `*/*`. The only difference is that WebFlux only has a Header strategy by default but it still should result in an empty list (unless the Header strategy is used directly).

I can see that getting back an empty list when the Accept header had `*/*` can be surprising but don't forget a `*/*` can be ignored also when another strategy comes after to provide another value, which is what SPR-10513 was all about. You could say we are consistent in ignoring a single `*/*`.

The bottom line is that that no Accept header or `*/*` means the same and this is why we accepted SPR-15013. Why does it make a difference for the purpose of this discussion? Neither one suggests use of HTML.

**Comment by bclozel** [comment]
As of Spring Boot 2.0, we've disabled by default suffix pattern matching (see #11105).

@rwinch given the latest comments in this issue, what are you thoughts?

**Comment by rwinch** [comment]
@rstoyanchev 

> The bottom line is that that no Accept header or */* means the same and this is why we accepted SPR-15013.

This seems to be an implementation detail and unrelated to the `ContentNegotiationStrategy` contract. From Spring Security's perspective, it does not know if an Accept header is being used to resolve the `MediaType`s. Spring Security needs to follow the `ContentNegotiationStrategy` contract which does not say anything about an empty result being the same as `MediaType.ALL`.

**Comment by rwinch** [comment]
> @rwinch should we improve the ContentNegotiationStrategy javadoc to underline that? This might be an implementation detail but it also reflects the behavior of most HTTP clients/servers out there.

If this is the contract of `ContentNegotiationStrategy`, then that would be a good idea. However, I think it is a shame that everyone consuming `ContentNegotiationStrategy` must provide an additional check if the Collection is empty, then to default it to be `MediaType.ALL`.

To me having the API setup this way is almost like returning a null value instead of an empty collection. Every user of such an API would be forced to check to see if the result is null and either default the collection to an empty collection. In the case of `ContentNegotiationStrategy` every user must perform a check to see if the Collection is empty and if so, default it to a collection of `MediaType.ALL`.

> I also don't get why WebFlux and MVC aren't considered the same way by Spring Security. Where does this inconsistency comes from?

The difference is that Spring Security does not use the `RequestedContentTypeResolver` for resolving media types in Spring Security. Instead it copied the code from `HeaderContentTypeResolver` which returns `MediaType.ALL` and does not return an empty collection.

The reason for this was packaging inconsistencies between MVC and WebFlux. For MVC you will notice that `ContentNegotiationStrategy` is in spring-web. However, for WebFlux `RequestedContentTypeResolver` is in the spring-webflux jar which Spring Security does not depend on. If I remember correctly (and based upon what the code reflects), when speaking with Rossen about this inconsistency the recommendation was to copy the very simple logic into our own class.

At some point we may provide an interface in Spring Security that allows adapting Spring WebFlux's `RequestedContentTypeResolver` to work with Spring Security, but we haven't had an immediate need for it yet (other things are taking priority).

**Comment by rstoyanchev** [comment]
@rwinch you're correct to point out an inconsistency between `ContentNegotiationStrategy` as a contract (and individual implementations) vs `ContentNegotiationManager`.  In retrospect, we could have said so explicitly in the contract, and then update the header strategy accordingly. In practice however, it really shouldn't matter, and the only reason we bother to check in `ContentNegotiationManager` is to give other strategies a chance, which as a consequence results in an empty list.

Why does the absence of media types affect Spring Security, i.e. why not treat that the same as `*/*` which is in effect the same alignment that was requested in SPR-10513?

**Comment by rstoyanchev** [comment]
Looks like our comments crossed. Sounds like your objection is purely based on an API inconsistency. FWIW this is typically not an issue outside `ContentNegotiationManager` where for content negotiation algorithm works more generally, not by checking individual media types.

**Comment by rwinch** [comment]
> Why does the absence of media types affect Spring Security, i.e. why not treat that the same as */* which is in effect the same alignment that was requested in SPR-10513?

We can certainly do this. However, I feel like this makes the `ContentNegotiationStrategy` more difficult to use for every consumer of the API. I mentioned that when our comments crossed, but to highlight it:

>> To me having the API setup this way is almost like returning a null value instead of an empty collection. Every user of such an API would be forced to check to see if the result is null and default the collection to an empty collection. In the case of ContentNegotiationStrategy every user must perform a check to see if the Collection is empty and if so, default it to a collection of MediaType.ALL.

This additional effort can be seen in [Spring MVC's](https://github.com/spring-projects/spring-framework/blob/f18d8c31e21528ac039c4fe865dcbbd9a0770489/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/condition/ProducesRequestCondition.java#L263) usage [of](https://github.com/spring-projects/spring-framework/blob/f18d8c31e21528ac039c4fe865dcbbd9a0770489/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/AbstractMessageConverterMethodProcessor.java#L369) the [API](https://github.com/spring-projects/spring-framework/blob/f18d8c31e21528ac039c4fe865dcbbd9a0770489/spring-webmvc/src/main/java/org/springframework/web/servlet/view/ContentNegotiatingViewResolver.java#L257-L259). Obviously this additional effort now need to be added to Spring Security.

PS: I didn't look at into it, but should [ReactiveTypeHandler](https://github.com/spring-projects/spring-framework/blob/f18d8c31e21528ac039c4fe865dcbbd9a0770489/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/ReactiveTypeHandler.java#L164) also default an empty collection with `MediaType.ALL`?

> Also do you actually reject requests with no media type even if */* is authorized?

It's not about if we reject a request, it is really about how we prompt a user for authentication.

If we are at the point where `ContentNegotiationStrategy` and `RequestedContentTypeResolver` are not going to change, then we can make the changes in Spring Security but I was hoping to make a change that would make it easier for anyone who consumes these two APIs.

**Comment by rstoyanchev** [comment]
Okay I misunderstood then what you are asking. Your [earlier comment](https://github.com/spring-projects/spring-boot/issues/10907#issuecomment-347933951) was along the lines of not loosing a `*/*` if that was present in the request. It sounds like what you're actually asking is for `ContentNegotiationStrategy` to never return an empty list.

I suppose we could refine the behavior since technically you have to be prepared to deal with both. However that would probably in master only.

**Comment by rwinch** [comment]
> Your earlier comment was along the lines of not loosing a */* if that was present in the request. It sounds like what you're actually asking is for ContentNegotiationStrategy to never return an empty list.

Sorry for the confusion. You are right about what I was originally asking. However, after our discussions it seems that never returning an empty list makes more sense.

> I suppose we could refine the behavior since technically a consumer of CNS has to be prepared to deal with both.

Thank you. Can we ensure that `RequestedContentTypeResolver` is updated in the same way?

> However we'd probably change this only in master.

This seems reasonable to me.

**Comment by rstoyanchev** [comment]
Yes, if you create a ticket. The idea is to always keep the two sides in sync where feasible whether you thought about asking or not :)

**Comment by bclozel** [comment]
@rstoyanchev @rwinch Thanks!
Could you report back here with the Security + Framework issues so that people can track those changes?

In the meantime, I'm closing this issue - feel free to reopen it if you feel some changes are required in Spring Boot.

**Comment by rwinch** [comment]
The Spring Framework ticket can be found at https://jira.spring.io/browse/SPR-16624
